### PR TITLE
feat: impl RangeArray based on DictionaryArray

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement the `RangeArray` as described in RFC: https://github.com/GreptimeTeam/greptimedb/blob/develop/docs/rfcs/2022-12-20-promql-in-rust/rfc.md#data-model

It's a basic type of the PromQL execute extension. To represent `Matrix` or `RangeVector` in Prometheus. See the document of `RangeArray` struct for a detailed explaination.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#596

### CODE REVIEW

File: src/promql/src/range_array.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    /// Arrow's API.
+    pub fn into_dict(self) -> DictionaryArray<Int64Type> {
+        self.array
+    }
+
+    fn check_ranges<R>(value_len: usize, ranges: R) -> Result<()>
+    where
+        R: IntoIterator<Item = (i32, i32)>,
+    {
+        for (offset, length) in ranges.into_iter() {
```
Comment (evenyag) [inline src/promql/src/range_array.rs:171]: Do we need to check whether `offset` or `length` is negative? And I just realize that why not use u64 and u32?

File: src/promql/src/range_array.rs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+                    offset..(offset + length)
+                })
+            })
+            .collect::<Vec<_>>();
+        writeln!(
+            f,
+            "DictionaryArray {{array: {:?} ranges: {:?}}}",
+            self.array.values(),
+            ranges
+        )
```
Comment (evenyag) [inline src/promql/src/range_array.rs:217]: Looks like you could use `f.debug_struct()`

File: src/promql/src/range_array.rs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+/// ┌───────┬───────┬───────┐      ┌───────┐
+/// │ (0,2) │ (1,2) │ (2,2) │ ─┐   │[A,B,C]│ values.slice(0,2)
+/// └───────┴───────┴───────┘  │   ├───────┤
+///  Values                    ├─► │[B,C,D]│ values.slice(1,2)
+/// ┌───┬───┬───┬───┬───┐      │   ├───────┤
+/// │ A │ B │ C │ D │ E │     ─┘   │[C,D,E]│ values.slice(2,2)
+/// └───┴───┴───┴───┴───┘          └───────┘
+/// ```
+pub struct RangeArray {
+    array: DictionaryArray<Int64Type>,
```
Comment (evenyag) [inline src/promql/src/range_array.rs:63]: Could we use `UInt64Type`?

Comment (waynexia) [reply src/promql/src/range_array.rs:63]: why bother?

**Review by evenyag — APPROVED** [review]
LGTM

File: src/promql/src/range_array.rs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        self.array
+    }
+
+    fn check_ranges<R>(value_len: usize, ranges: R) -> Result<()>
+    where
+        R: IntoIterator<Item = (u32, u32)>,
+    {
+        for (offset, length) in ranges.into_iter() {
+            ensure!(
+                offset + length <= value_len as _,
```
Comment (v0y4g3r) [inline src/promql/src/range_array.rs:168]: Better to cast offset and length to `usize` like line 150, after all `(u32::MAX, 1)` could be a valid range.
```suggestion
                offset as usize + length as usize <= value_len,
```

File: src/promql/src/range_array.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+/// in [RangeArray] to represent a "range":
+///
+/// ```text
+/// 63            32│31             0
+/// ┌───────────────┼───────────────┐
+/// │  offset (i32) │  length (i32) │
+/// └───────────────┼───────────────┘
+/// ```
+///
+/// Then the [DictionaryArray] can be expand to serval ranges like this:
```
Comment (v0y4g3r) [inline src/promql/src/range_array.rs:50]:
```suggestion
/// Then the [DictionaryArray] can be expanded to serveral ranges like this:
```

File: src/promql/src/range_array.rs (hunk 4 above)
Comment (waynexia) [reply src/promql/src/range_array.rs:168]: 👍

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**